### PR TITLE
Compilation fixes for Coq 8.3

### PR DESCRIPTION
**Pull request by herbelin**
In the gallineo-camelo-archeology serial, one can observe that 8.3 fails to compile with ocaml 4.04.0 due to the conflict with OCaml files `matching.ml` and `lexer.ml`. I modified the `configure` to inform about this incompatibility.

Worse than that, the same conflict exists in principle with OCaml 4.03.2 but it does not affect the compilation with 4.03.2, suspectingly because the criticality of this conflict is dependent on the order in which the conflicting names arrives in some table.

### CODE REVIEW

**Comment by gares** [comment]
I guess you made a mistake in git, this PR includes 250+ commits while you message talks about a fix to the configure...

**Comment by ejgallego** [comment]
The target branch in github was wrong, FTFY.

File: configure [hunk 1]
```diff
@@ -427,16 +427,16 @@ esac
 CAMLVERSION=`"$bytecamlc" -version`
 
 case $CAMLVERSION in
-    1.*|2.*|3.00|3.01|3.02|3.03|3.03alpha|3.04|3.05beta|3.05|3.06|3.07*|3.08*|3.09*)
+    1.*|2.*|3.00|3.01|3.02|3.03|3.03alpha|3.04|3.05beta|3.05|3.06|3.07*|3.08*|3.09*|4.0[4-9]*|4.[1-9]*|5.*)
```
Comment (JasonGross) [inline configure]: Should the `5` case be `[5-9].*` or `[5-9].*|[1-9][0-9]*`?

**Comment by herbelin** [comment]
@gares, sorry for the noise.

Sorry also for what happens to have been a wip. I realized afterwards that there is a simple way to workaround the conflicts with OCaml's own file names (and this could be applicable to the case of `profile.ml` mentioned in #6325), namely to deactivate warning 31. Since we didn't have problems for years without caring about this warning, I speculate that it is safe to continue to work without it (But anyway, in the case of `profile.ml`, there is already the [profile commit](afee1cf8a74add) of #6038, so this does not matter anymore.)

So, back to old 8.3, it finally seems to compile fine with all versions of OCaml up to 4.06.0.

@jasongross: the patch is made simpler, so what you mentioned does not matter anymore.

**Comment by JasonGross** [comment]
@herbelin Did you mean to close this?  (I guess it automatically closed when you pushed the tip of v8.3 to the PR?)

**Comment by herbelin** [comment]
@JasonGross: Indeed, I pushed to the wrong branch. Should be ok now. Thanks.

File: configure [hunk 2]
```diff
@@ -436,8 +438,23 @@ case $CAMLVERSION in
 	    echo "          Configuration script failed!"
 	    exit 1
 	fi;;
+    3.10*|3.11*|3.12*)
+	CAMLP4COMPAT="-loc loc" 
+	echo "You have Objective-Caml $CAMLVERSION. Good!";;
+    4.02.0)
+	CAMLP4COMPAT="-loc loc"
+	echo "Warning: OCaml 4.02.0 suffers from a bug inducing"
+	echo "very slow compilation times";;
```
Comment (williammpratt) [inline configure]: Very slow has no meaning. Does it take 10 hours on a modern laptop to compile it, 100, or 10000?

**Pull request merged**